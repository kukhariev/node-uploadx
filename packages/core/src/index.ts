export * from './handlers';
export * from './storages';
export * from './utils';
export * from './types';
