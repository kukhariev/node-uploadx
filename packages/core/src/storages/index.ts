export * from './disk-storage';
export * from './file';
export * from './storage';
export * from './local-meta-storage';
export * from './meta-storage';
export * from './config';
