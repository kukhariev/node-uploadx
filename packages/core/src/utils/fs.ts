import { createWriteStream, promises as fsp, WriteStream } from 'fs';
import { dirname, posix } from 'path';

/**
 * Ensures that the directory exists
 * @param dir
 */
export async function ensureDir(dir: string): Promise<void> {
  await fsp.mkdir(dir, { recursive: true });
}

/**
 * Ensures that the file exists and returns it size
 * @param path
 * @param overwrite Force creating new empty file
 * @returns file size
 */
export async function ensureFile(path: string, overwrite = false): Promise<number> {
  await fsp.mkdir(dirname(path), { recursive: true });
  await (await fsp.open(path, overwrite ? 'w' : 'a')).close();
  return (await fsp.stat(path)).size;
}

/**
 * Returns file WriteStream for data appending.
 * @param path
 * @param start
 */
export function getWriteStream(path: string, start: number): WriteStream {
  if (path && start >= 0) {
    return createWriteStream(path, { flags: 'r+', start });
  }
  throw new Error('getWriteStream: invalid parameters!');
}

/**
 * Return file paths that begin with the prefix
 * @param prefix
 */
export function getFiles(prefix: string): Promise<string[]> {
  const prefix_ = prefix.replace(/\\/g, '/');
  const _getFiles = async (current: string): Promise<string[]> => {
    try {
      if ((await fsp.stat(current)).isFile()) return [current];
    } catch {
      return _getFiles(dirname(current));
    }
    const dirents = await fsp.readdir(current, { withFileTypes: true });
    const files = await Promise.all(
      dirents.map(async dirent => {
        const path = posix.join(current, dirent.name);
        return path.startsWith(prefix_) ? (dirent.isDirectory() ? _getFiles(path) : path) : null;
      })
    );
    return Array.prototype
      .concat(...files)
      .filter(Boolean)
      .sort() as string[];
  };
  return _getFiles(prefix_);
}

export { fsp };
