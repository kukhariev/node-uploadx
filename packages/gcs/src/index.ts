export * from './gcs-storage';
export * from './gcs-meta-storage';
export * from './constants';
