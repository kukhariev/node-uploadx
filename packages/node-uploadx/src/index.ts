export * from '@uploadx/core';
export * from '@uploadx/s3';
export * from '@uploadx/gcs';
