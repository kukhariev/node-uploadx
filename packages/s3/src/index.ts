export * from './s3-storage';
export * from './s3-meta-storage';
export * from './aws-error';
