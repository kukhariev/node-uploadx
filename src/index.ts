export * from './core';
export * from './core/utils';
export * from './Tus';
export * from './Uploadx';
export * from './Multipart';
