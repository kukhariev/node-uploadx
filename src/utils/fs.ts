import { promises as fsp } from 'fs';
import { dirname, resolve } from 'path';

export async function ensureDir(dir: string): Promise<void> {
  await fsp.mkdir(dir, { recursive: true });
}

export async function ensureFile(path: string, overwrite = false): Promise<number> {
  await fsp.mkdir(dirname(path), { recursive: true });
  await (await fsp.open(path, overwrite ? 'w' : 'a')).close();
  const { size } = await fsp.stat(path);
  return size;
}

/**
 * Resolve the Promise with file size or `-1` on error
 *
 * @param path
 */
export async function getFileSize(path: string): Promise<number> {
  try {
    return (await fsp.stat(path)).size;
  } catch {
    return -1;
  }
}
export async function getFiles(prefix: string): Promise<string[]> {
  try {
    if ((await fsp.stat(prefix)).isFile()) return [prefix];
  } catch {
    return [];
  }
  const dirents = await fsp.readdir(prefix, { withFileTypes: true });

  const files = await Promise.all(
    dirents.map(async dirent => {
      const path = resolve(prefix, dirent.name);
      return dirent.isDirectory() ? getFiles(path) : path;
    })
  );
  return Array.prototype.concat(...files);
}
export { fsp };
