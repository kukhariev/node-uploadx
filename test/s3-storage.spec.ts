import { S3 } from 'aws-sdk';
import { createReadStream } from 'fs';
import { FilePart, S3File, S3Storage } from '../src';
import { storageOptions } from './fixtures';
import { filename, metafile, srcpath, testfile } from './fixtures/testfile';

const mockHeadBucket = jest.fn();
const mockCreateMultipartUpload = jest.fn(() => ({
  promise() {
    return Promise.resolve({ UploadId: '123456789' });
  }
}));
const mockPutObject = jest.fn(() => ({
  promise() {
    return Promise.resolve();
  }
}));
const mockListObjectsV2 = jest.fn(() => ({
  promise() {
    return Promise.resolve<S3.ListObjectsV2Output>({
      Contents: [
        { Key: 'already.uploaded', LastModified: new Date() },
        { Key: metafile, LastModified: new Date() }
      ]
    });
  }
}));
const mockDeleteObject = jest.fn(() => ({
  promise() {
    return Promise.resolve();
  }
}));
const mockAbortMultipartUpload = jest.fn(() => ({
  promise() {
    return Promise.resolve();
  }
}));
const mockUploadPart = jest.fn(() => ({
  promise() {
    return Promise.resolve();
  }
}));
const mockCompleteMultipartUpload = jest.fn(() => ({
  promise() {
    return Promise.resolve({ Location: '' });
  }
}));
const mockHeadObject = jest.fn(() => ({
  promise() {
    return Promise.resolve({
      Metadata: { metadata: encodeURIComponent(JSON.stringify(testfile)) }
    });
  }
}));
const mockListParts = jest.fn(() => ({
  promise() {
    return Promise.resolve({
      Metadata: { Parts: [{ PartNumber: 1, Size: testfile.size }] }
    });
  }
}));

jest.mock('aws-sdk', () => ({
  S3: jest.fn(
    (): Partial<S3> => ({
      abortMultipartUpload: mockAbortMultipartUpload as any,
      completeMultipartUpload: mockCompleteMultipartUpload as any,
      createMultipartUpload: mockCreateMultipartUpload as any,
      deleteObject: mockDeleteObject as any,
      headBucket: mockHeadBucket as any,
      headObject: mockHeadObject as any,
      listObjectsV2: mockListObjectsV2 as any,
      listParts: mockListParts as any,
      putObject: mockPutObject as any,
      uploadPart: mockUploadPart as any
    })
  )
}));

describe('S3Storage', () => {
  const options = { ...storageOptions };
  testfile.name = filename;

  let file: S3File;
  let storage: S3Storage;
  beforeEach(async () => {
    storage = new S3Storage(options);
    file = {
      ...testfile,
      UploadId: '123456789',
      Parts: []
    };
  });
  describe('.create()', () => {
    it('should request api and set status and uri', async () => {
      mockHeadObject.mockImplementationOnce(() => ({
        promise() {
          return Promise.reject({ statusCode: 404 });
        }
      }));
      file = await storage.create({} as any, testfile);
      expect(file.name).toEqual(filename);
      expect(file.status).toBe('created');
      expect(file).toMatchObject({
        ...testfile,
        UploadId: expect.any(String),
        Parts: expect.any(Array)
      });
    });

    it('should return existing file', async () => {
      storage['cache'] = { [filename]: { ...file } };
      const existing = await storage.create({} as any, testfile);
      expect(existing.UploadId).toEqual(file.UploadId);
    });
  });
  describe('.update()', () => {
    it('should update changen metadata keys', async () => {
      file = await storage.update(filename, { metadata: { name: 'newname.mp4' } } as any);
      expect(file.metadata.name).toBe('newname.mp4');
      expect(file.metadata.mimeType).toBe(testfile.metadata.mimeType);
    });

    it('should reject if not found', async () => {
      expect.assertions(1);
      mockHeadObject.mockImplementationOnce(() => ({
        promise() {
          return Promise.reject({ statusCode: 404 });
        }
      }));
      await expect(
        storage.update(filename, { metadata: { name: 'newname.mp4' } } as any)
      ).rejects.toHaveProperty('statusCode', 404);
    });
  });
  describe('.get()', () => {
    it('should return all user files', async () => {
      const files = await storage.get(testfile.userId);
      expect(files).toEqual(expect.any(Array));
      expect(files.length).toEqual(1);
      expect(files[0]).toMatchObject({ name: filename });
    });
  });
  describe('.write()', () => {
    it('should request api and set status and bytesWritten', async () => {
      const part: FilePart = {
        name: filename,
        body: createReadStream(srcpath),
        start: 0,
        contentLength: testfile.size
      };
      const res = await storage.write(part);
      expect(res.status).toBe('completed');
      expect(res.bytesWritten).toBe(testfile.size);
    });

    it('should request api and set status and bytesWritten on resume', async () => {
      const part: FilePart = {
        name: filename,
        contentLength: 0
      };
      const res = await storage.write(part);
      expect(res.status).toBe('part');
      expect(res.bytesWritten).toBe(0);
    });
  });
  describe('delete()', () => {
    it('should set status', async () => {
      const [deleted] = await storage.delete(filename);
      expect(deleted.status).toBe('deleted');
    });
  });
});
