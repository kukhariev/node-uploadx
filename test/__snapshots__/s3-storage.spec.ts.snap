// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3Storage .create() should handle existing 1`] = `
{
  "UploadId": "987654321",
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "userId/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
  "userId": "userId",
}
`;

exports[`S3Storage .create() should request api and set status and UploadId 1`] = `
S3File {
  "UploadId": "123456789",
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "userId/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
  "userId": "userId",
}
`;

exports[`S3Storage .create() should send error on invalid s3 response 1`] = `
{
  "detail": "s3 create multipart upload error",
  "message": "FileError",
  "name": "UploadxError",
  "uploadxErrorCode": "FileError",
}
`;

exports[`S3Storage .write() should request api and set status and bytesWritten 1`] = `
{
  "UploadId": "987654321",
  "bytesWritten": 64,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "userId/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "completed",
  "uri": "/1234",
  "userId": "userId",
}
`;

exports[`S3Storage .write() should request api and set status and bytesWritten on resume 1`] = `
{
  "Parts": [],
  "UploadId": "987654321",
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "userId/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "part",
  "userId": "userId",
}
`;

exports[`S3Storage normalizeError not s3 error 1`] = `
{
  "code": "GenericUploadxError",
  "message": "Generic Uploadx Error",
  "statusCode": 500,
}
`;

exports[`S3Storage normalizeError s3 error 1`] = `
{
  "code": "SomeServiceException",
  "message": "SomeServiceException",
  "name": "SomeError",
  "statusCode": 400,
}
`;
